File: packages/jskit/src/scope/options.ts
/**
 * @fileoverview The options an analysis runs with.
 */

/**
 * How a program should be interpreted while its scopes are built.
 */
export interface AnalyzeOptions {
	/** Whether the program is a script, an ES module, or a CommonJS module. */
	sourceType?: "script" | "module" | "commonjs";

	/**
	 * Whether TypeScript syntax carries meaning.
	 *
	 * A valid JavaScript program contains no TypeScript nodes, so the type
	 * rules have nothing to fire on either way. What this actually decides is
	 * the one construct the two reference implementations disagree about:
	 * `export { a }` names both a value and a type under `"ts"`, and only a
	 * value under `"js"`, which is what `eslint-scope` reports.
	 */
	dialect?: "js" | "ts";

	/** Whether a JSX identifier counts as a reference. */
	jsx?: boolean;

	/** Whether strict mode applies without a directive saying so. */
	impliedStrict?: boolean;

	/**
	 * Whether an extra function scope wraps the program, which is what makes
	 * top-level `return` legal. Implied by `sourceType: "commonjs"`.
	 */
	globalReturn?: boolean;

	/**
	 * Whether a direct call to `eval` should be ignored rather than making its
	 * enclosing scopes dynamic.
	 */
	ignoreEval?: boolean;

	/**
	 * Names to declare in the global scope, which is how a host's built-ins
	 * are supplied. References waiting for them resolve once they exist.
	 */
	globals?: Iterable<string>;

	/**
	 * The name a JSX element compiles a call to, referenced once per element
	 * when set. `@typescript-eslint/scope-manager` defaults this to `"React"`;
	 * here it defaults to `null`, because `eslint-scope` creates no such
	 * reference and `eslint-scope` is the tiebreaker.
	 */
	jsxPragma?: string | null;

	/** The name a JSX fragment compiles a call to, referenced when set. */
	jsxFragmentName?: string | null;
}

/**
 * The same options with every default filled in.
 */
export interface ResolvedOptions {
	sourceType: "script" | "module" | "commonjs";
	dialect: "js" | "ts";
	jsx: boolean;
	impliedStrict: boolean;
	globalReturn: boolean;
	ignoreEval: boolean;
	globals: Iterable<string> | null;
	jsxPragma: string | null;
	jsxFragmentName: string | null;
}

/**
 * Fills in the defaults an analysis was not given.
 * @param options The options the caller supplied.
 * @returns The options with every field present.
 */
export function resolveOptions(options: AnalyzeOptions = {}): ResolvedOptions {
	return {
		sourceType: options.sourceType ?? "module",
		dialect: options.dialect ?? "ts",
		jsx: options.jsx ?? true,
		impliedStrict: options.impliedStrict ?? false,
		globalReturn: options.globalReturn ?? false,
		ignoreEval: options.ignoreEval ?? false,
		globals: options.globals ?? null,
		jsxPragma: options.jsxPragma ?? null,
		jsxFragmentName: options.jsxFragmentName ?? null,
	};
}

File: packages/jskit-native/crates/jskit-core/src/scope/options.rs
//! The options an analysis runs with.
//!
//! Port of `packages/jskit/src/scope/options.ts`, with every default filled
//! in. Names are UTF-16 code units, because they are compared against and
//! interned beside names sliced from the source text.

use super::buffer::*;

/// Whether the program is a script, an ES module, or a CommonJS module.
#[derive(Clone, Copy, PartialEq, Eq, Debug, Default)]
pub enum ScopeSourceType {
    Script,
    #[default]
    Module,
    CommonJs,
}

/// The analysis options with every default filled in.
#[derive(Clone, Default)]
pub struct ResolvedOptions {
    pub source_type: ScopeSourceType,

    /// Whether TypeScript syntax carries meaning (`dialect: "ts"`).
    pub dialect_ts: bool,

    /// Whether a JSX identifier counts as a reference.
    pub jsx: bool,

    /// Whether strict mode applies without a directive saying so.
    pub implied_strict: bool,

    /// Whether an extra function scope wraps the program.
    pub global_return: bool,

    /// Whether a direct call to `eval` should be ignored.
    pub ignore_eval: bool,

    /// Names to declare in the global scope, or `None`.
    pub globals: Option<Vec<Vec<u16>>>,

    /// The name a JSX element compiles a call to, or `None`.
    pub jsx_pragma: Option<Vec<u16>>,

    /// The name a JSX fragment compiles a call to, or `None`.
    pub jsx_fragment_name: Option<Vec<u16>>,
}

impl ResolvedOptions {
    /// The TypeScript defaults: `sourceType: "module"`, `dialect: "ts"`,
    /// `jsx: true`, everything else off.
    pub fn defaults() -> Self {
        ResolvedOptions {
            source_type: ScopeSourceType::Module,
            dialect_ts: true,
            jsx: true,
            implied_strict: false,
            global_return: false,
            ignore_eval: false,
            globals: None,
            jsx_pragma: None,
            jsx_fragment_name: None,
        }
    }

    /// The `SCOPE_H_OPTIONS` header word.
    pub fn encoded(&self) -> u32 {
        (match self.source_type {
            ScopeSourceType::Module => OPT_SOURCE_TYPE_MODULE,
            ScopeSourceType::CommonJs => OPT_SOURCE_TYPE_COMMONJS,
            ScopeSourceType::Script => OPT_SOURCE_TYPE_SCRIPT,
        }) | (if self.dialect_ts { OPT_DIALECT_TS } else { 0 })
            | (if self.jsx { OPT_JSX } else { 0 })
            | (if self.implied_strict { OPT_IMPLIED_STRICT } else { 0 })
            | (if self.global_return { OPT_GLOBAL_RETURN } else { 0 })
            | (if self.ignore_eval { OPT_IGNORE_EVAL } else { 0 })
    }
}
